R/bivariate.density.R

```diff
@@ -293,24 +293,30 @@ bivariate.density <- function(pp,h0,hp=NULL,adapt=FALSE,resolution=128,gamma.sca
 		notin <- !inside.owin(x=evalxy[,1],y=evalxy[,2],w=W)
 		surf <- rep(NA,nrow(evalxy))
 		ef <- NULL
-		
+
+		# Bivariate kernel function centered at 0,0 with isotropic bandwidth h
+		kernel2d <- function(x, y, h) {
+		  1/(2*pi*h*h)*exp(-0.5*(x*x+y*y)/(h*h))
+		}
+
+		# Approximation of definite integral of f over a grid with gridsize dx x dy via summation
+		dintegral <- function(f, dx, dy) {
+		  sum(f)*dx*dy
+		}
+
 		if(edge=="uniform"){
 			qhz <- rep(NA,resolution^2)
 			if(verbose) pb <- txtProgressBar(0,nrow(evalxy))
 			for(i in 1:nrow(evalxy)){
 				ht <- h.hypo.mat[which(pilot.density$yrow==evalxy[i,2]),which(pilot.density$xcol==evalxy[i,1])]
 				if(is.na(ht)) next
```

Comment:
Notice here that you're breaking the loop before you compute ivals (and thus surf[i]) for these observations.

This is probably OK, as my guess is you throw them away anyway as they're out of region?

I suggest we break the loops into two, so you do the edge correction first for each case, then the common bit (computing ivals/surf) you do once instead of repeating it. i.e. make everything like the diggle case, then pull the common bit out.